### PR TITLE
feat(appmesh): change VirtualService provider to a union-like class

### LINKED ISSUE #9490: [@aws-cdk/aws-appmesh] AppMesh Stable Changes
### AppMesh API Revisions

1. Currently `VirtualNode` defaults the value of the logging file path parameter to `/dev/stdout` and is not configurable through the construct's props. Remove this default and allow customers to configure log location for their specific use case.

2. Change `addListeners` method signature to not be variadic since `VirtualNode` only has one listener. Depending on https://github.com/aws/aws-app-mesh-roadmap/issues/120 we may or may not actually want to change this API.

3. Change model so that `VirtualService` always uses mesh from either router or node and requires mesh property to exist on `VirtualRouter` and `VirtualNode` for importing. `VirtualNode` and `VirtualRouter` always have to be in same mesh as `VirtualService`.
--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by dfezzie**
Fixes #9490

BREAKING CHANGE: the properties virtualRouter and virtualNode of VirtualServiceProps have been replaced with the union-like class VirtualServiceProvider 
* **appmesh**: the method `addVirtualService` has been removed from `IMesh`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:58]: 2 changes here:

1. Let's move this property to `VirtualServiceProps` below.
2. Let's remove the `mesh` property from `VirtualServiceProps` (no reason to provide it redundantly when a `VirtualRouter` or `VirtualNode` were provided).

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:148]:
```suggestion
        provider: provider
          ? {
            virtualNode: provider?.virtualNodeProvider,
            virtualRouter: provider?.virtualRouterProvider,
          }
          : undefined,
```

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:199]: Let's add a third, required, property here, `mesh: IMesh`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:243]:
```suggestion
      virtualNodeProvider: this.virtualNode
        ? {
          virtualNodeName: this.virtualNode.virtualNodeName,
        }
        : undefined,
      virtualRouterProvider: this.virtualRouter
        ? {
          virtualRouterName: this.virtualRouter.virtualRouterName,
        }
        : undefined,
```

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:219]: Let's add a third constructor here, `mesh(mesh: IMesh): VirtualServiceProvider`.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks great @dfezzie ! A few comments here, mainly related to this comment that I made: https://github.com/aws/aws-cdk/issues/9490#issuecomment-742749875 .

Please also add `addVirtualService()` methods to both `IVirtualNode` and `IVirtualRouter`.

Also, I'm fairly sure that's the last point that's left of #9490, so I would add `Fixes #9490` to the PR description 🙂.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:221]: What is the implication of this? that there is no provider?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:58]: 1. Do we not need the provider in the base props? How else will we define what type of resource backs the virtual service? 

2. Since the provider is optional, I don't think we can remove the `mesh` property. We still need to know what mesh the VS is created as a part of. We can enforce the mesh properties on the VS and Provider do match.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:221]: I think so. If you want, you can call the method `none` instead: `none(mes: IMesh): VirtualServiceProvider`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:58]: 1. We do not, because the provider will be evident from the resource that the `addService` method will be on. For example, for `VirtualNode`, it will be:

```ts
public addService(props: VirtualServiceBaseProps): IVirtualService: {
  return new VirtualService(this, 'VirtualService', {
    ...props,
    provider: ServiceProvider.virtualNode(this),
  });
}
```

And similarly for `Mesh` and `VirtualRouter`.

2. The `provider` should be required, of course 🙂.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:58]: > Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!

I ended up getting sidetracked and haven't addressed all the feedback quite yet

**Comment by dfezzie** [comment]
> Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!

I ended up getting sidetracked and haven't addressed all the feedback quite yet

**Comment by skinny85** [comment]
> > Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!
> 
> I ended up getting sidetracked and haven't addressed all the feedback quite yet

No worries, it's all good 🙂

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:231]: Thinking on this more, our API spec does not require you to define a provider but the resource would not provide any usability in that state. Providing guardrails here ensures that customers have configured their mesh properly.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:204]: This is present. May be a stale comment?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:58]: So I've provided a comment on https://github.com/aws/aws-cdk/issues/9490 that argues we should not have an `addService` method on the `IVirtualNode` and `IVirtualRouter` classes. Will wait on your response before I do anything here.

**Comment by dfezzie** [comment]
Sorry for the long delay. Holidays snuck up on me and then this PR fell off my radar 😅 

I've provided a comment on https://github.com/aws/aws-cdk/issues/9490 about the changes requested. Depending on your response there, we may be able to call this PR good to go.

**Review by skinny85 — DISMISSED** [review]
@dfezzie I have no problems with not having `addService()` methods in `IVirtualNode` and `IVirtualRouter`. If that's the case, I think you might want to get rid of the existing `addVirtualService()` method in `IMesh` as well...?

Having said that, I do feel strongly that requiring the user to provide the Mesh twice when creating a Virtual Service with a provider is a mistake (as both a VirtualNode and VirtualRouter already have a Mesh). It's just poor usability.

If you're very opposed to my idea of having a third `none()` VirtualServiceProvider, let's at least make `IMesh` an optional argument when creating a `VirtualService`, and we will not not require it when a `provider` has been given. How does that sound?

**Comment by dfezzie** [comment]
I think we should make `provider` a required argument altogether, get rid of the `IMesh` argument and just just get it from the provider. How does that sound?

**Comment by dfezzie** [comment]
I've added the `none()` method to the class. Gives the customer flexibility to define a Virtual Service without having to provide a provider out of the gate. 

I'm having issues deploying the integ test to my own account, seeing 

```
1 validation error detected: Value 'local' at 'domainValidationOptions.1.member.validationDomain' failed to satisfy constraint: Member must satisfy regular expression pattern: ^(\*\.)?(((?!-)[A-Za-z0-9-]{0,62}[A-Za-z0-9])\.)+((?!-)[A-Za-z0-9-]{1,62}[A-Za-z0-9])$ (Service: AWSCertificateManager
```

Not sure if there's a bug with the deprecated props in `acm.Certificate` or something weird with my setup. I'll dig into it a bit more tomorrow if the build here doesn't pass

**Comment by skinny85** [comment]
@dfezzie for deploying the `integ.mesh.ts` integ test, maybe reach out to @alexbrjo ? I think he was the last person to modify them.

**Comment by dfezzie** [comment]
I had to remove ACMPCA and ACM from the integ tests as App Mesh does validate the existence of the resources, and as a result you cannot just import from a fake ARN. I think that should fix the last of the issues with the build.

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:119]: Not sure when I removed this. Will add back.

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/mesh.ts:92]: Sorry, I don't understand these changes 😕.

The whole point of having `addVirtualService()` on Mesh is so that you don't have to provide the Mesh again. But with your changes, you can do something like:

```ts
const mesh = new appmesh.Mesh(...);
mesh.addVirtualService({
  virtualServiceProvider.none(mesh),
});
```

which just doesn't make sense to me.

I would revert all of these changes here, and this method should be:

```suggestion
  public addVirtualService(id: string, props: VirtualServiceBaseProps = {}): VirtualService {
    return new VirtualService(this, id, {
      ...props,
      virtualServiceProvider: VirtualServiceProvider.none(this),
    });
  }
```

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:84]: Like I wrote above in `mesh.ts`, I would revert all of these changes, keep `VirtualServiceBaseProps`, and change `readonly mesh: IMesh` in `VirtualServiceProps` to `readonly virtualServiceProvider: VirtualServiceProvider`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:128]: This check doesn't make sense to me... the code is

```ts
    this.mesh = provider.mesh;
    if (provider.mesh != this.mesh) {
      // ...
    }
```

So the above `if` will never have its condition be true.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:231]: Let's not expose this (`VirtualServiceProviderConfig` already contains a `readonly mesh: IMesh` property).

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:242]: Given the above comment, let's also make this field `private readonly`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:250]: I think this is dead code (this `if` will never have its condition as true).

Let's just do this instead:

```ts
     const providedMesh = this.virtualNode?.mesh ?? this.virtualRouter?.mesh ?? mesh!;
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:39]: Empty line before this line.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:55]: Empty line before this line.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks almost perfect @dfezzie! Just a tiny change in `Mesh`, and this will be ready to be shipped.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:250]: Didn't know about the `!`. Neat!

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/mesh.ts:92]: I guess I was thinking of it more as a utility method, a quick way to add a new Virtual Service to your mesh, similar to how we add other resources like Virtual Nodes. If you cannot specify a provider, then essentially the method is useless as it will never produce a usable Virtual Service. If you created a VS as a placeholder and then went to add a provider you would have to remove it from the `addVirtualService` method into a `new VirtualService(...)` call. How it is here, you could have `addVirtualService('VirtualService')` in one synthesized stack which uses the `none` method as the default. Later, you could come along and add a different provider if you so choose. 

We can remove the method all together if you feel it does not fit well.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:128]: Ah good call. Fixed

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:231]: We need to access this in the Virtual Service Constructor class. The `VirtualServiceProps` only has access to the provider class: `readonly virtualServiceProvider: VirtualServiceProvider;` so we need to access the mesh there to properly create the associated cloudformation.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:84]: Done!

**Comment by dfezzie** [comment]
Addressed a few comments, waiting for a response before making any final updates :+1:

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/mesh.ts:92]: > I guess I was thinking of it more as a utility method, a quick way to add a new Virtual Service to your mesh

Right, but if you want to specify _any_ property of the created Service, then it doesn't really save anything, right?

```ts
mesh.addVirtualService('Service', {
  virtualServiceName: 'MyService',
  provider: VirtualServiceProvider.none(mesh),
});
```

Doesn't save anything over

```ts
new VirtualService(this, 'Service', {
  virtualServiceName: 'MyService',
  provider: VirtualServiceProvider.none(mesh),
});
```

I think it doesn't make too much sense to over-emphasize the "add a VirtualService to the Mesh with only default properties" use case - it seems to just lead to an awkward API.

> , similar to how we add other resources like Virtual Nodes.

Notice that all other methods int the AppMesh module like `addVirtualNode()`, `addVirtualGateway()`, etc. do _not_ allow you to specify the main owning resource, like the Mesh, again.

I don't think this inconsistency is worth it. I would either not allow passing `virtualServiceProvider` in `addVirtualService()`, or get rid of this method completely, whichever you prefer (if you go with the later, don't forget to add a BREAKING CHANGE note to the PR description).

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:232]: Not sure I understand, sorry.

In the constructor of `VirtualService`, you should call `bind()` on the `virtualServiceProvider` you get, and take the Mesh from there.

There's no point in having 2 separate places where a Mesh is surfaced by.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:232]: In fact, this is exactly how the code looks today:

```ts
    const provider = props.virtualServiceProvider?.bind(this);
    this.mesh = provider.mesh;
```

So I'm pretty sure this property is unused.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:232]: https://github.com/aws/aws-cdk/pull/11978/files?file-filters%5B%5D=.ts#diff-c7421c5311bdc768c94ddba1234eb002ff0c1d2d4a5e4e5d31724ea762ec7948R130-R131

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/mesh.ts:92]: For now, let's remove it. You're right that there is no real difference between the two method calls

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:232]: For some reason I was thinking the bind class was returning an instance of the `VirtualServiceProvider` class 🤦  

Fixed in this revision. Good call

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:43]: Moved these methods from the `test.mesh.ts` file. Did not modify them beyond changing `mesh.addVirtualService(...)` to `new appmesh.VirtualService(...)`

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:68]: Now that you've removed `addVirtualService()` from `IMesh`, I'm pretty sure the `VirtualServiceBaseProps` is not used anywhere. If I'm right, get rid of the it in favor of `VirtualServiceProps`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:130]: I'm 99% sure `provider` is required here, and so the `?.` can be turned to `.`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:140]: Same here - `provider` is never `undefined` here, so no need for the `?.`.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:137]: Gotta say, the name `provider` for this local variable is a little confusing... how about `providerConfig` instead?

**Review by skinny85 — DISMISSED** [review]
Looks great! Couple of cosmetic changes before we merge this in.

**Review by skinny85 — APPROVED** [review]
Looks great @dfezzie!

**Pull request merged**
